webpack.config.js

```diff
@@ -0,0 +1,245 @@
+			chunkFilename: '[name].chunk.[chunkhash:5].js',
+			path: path.join(__dirname, 'build'),
+			publicPath: '/'
+		},
+		resolve: {
+			extensions: ['.ts', '.tsx', '.js', '.jsx', '.scss', '.css'],
+			alias: {
+				'app-entry-point': path.join(__dirname, 'src/index'),
+				style: path.join(__dirname, 'src/style'),
+				preact$: path.join(__dirname, 'node_modules/preact/dist/preact.js')
+			}
+		},
+		resolveLoader: {
+			alias: {
+				async: path.join(__dirname, 'config/async-component-loader')
+			}
+		},
+		module: {
+			rules: [
+				{
+					test: /\.tsx?$/,
+					enforce: 'pre',
+					// loader: 'awesome-typescript-loader',
+					loader: 'ts-loader',
+					exclude: nodeModules
+				},
+				{
+					test: /\.(tsx?|jsx?)$/,
+					loader: 'babel-loader',
+					options: babelRc
+				},
+				{
+					test: /\.(scss|sass)$/,
+					loader: 'sass-loader',
+					enforce: 'pre',
+					options: {
+						sourceMap: true,
+						includePaths: [nodeModules]
+					}
+				},
+				{
+					test: /\.(scss|sass|css)$/,
+					include: componentStyleDirs,
+					use: [
+						isProd ? MiniCssExtractPlugin.loader : 'style-loader',
+						{
+							// loader: 'typings-for-css-modules-loader?modules&localIdentName=[local]__[hash:base64:5]&importLoaders=1'+(isProd ? '&sourceMap' : '')
+							// loader: 'css-loader',
+							loader: 'typings-for-css-modules-loader',
+							options: {
+								modules: true,
+								localIdentName: '[local]__[hash:base64:5]',
+								namedExport: true,
+								camelCase: true,
+								importLoaders: 1,
+								sourceMap: isProd,
+								sass: true
+							}
+						}
+					]
+					// loader: ExtractTextPlugin.extract({
+					// 	fallback: 'style-loader',
+					// 	use: [{
+					// 		loader: 'css-loader',
+					// 		options: {
+					// 			modules: true,
+					// 			localIdentName: '[local]__[hash:base64:5]',
+					// 			importLoaders: 1,
+					// 			sourceMap: isProd
+					// 		}
+					// 	}]
+					// })
+				},
+				{
+					test: /\.(scss|sass|css)$/,
+					exclude: componentStyleDirs,
+					use: [
+						isProd ? MiniCssExtractPlugin.loader : 'style-loader',
+						{
+							loader: 'css-loader',
+							options: {
+								importLoaders: 1,
+								sourceMap: isProd
+							}
+						}
+					]
+					// loader: ExtractTextPlugin.extract({
+					// 	fallback: 'style-loader',
+					// 	use: [{
+					// 		loader: 'css-loader',
+					// 		options: {
+					// 			importLoaders: 1,
+					// 			sourceMap: isProd
+					// 		}
+					// 	}]
+					// })
+				}
+			]
+		},
+		plugins: [
+			new ProgressBarPlugin({
+				format: '\u001b[90m\u001b[44mBuild\u001b[49m\u001b[39m [:bar] \u001b[32m\u001b[1m:percent\u001b[22m\u001b[39m (:elapseds) \u001b[2m:msg\u001b[22m',
+				renderThrottle: 100,
+				summary: false,
+				clear: true
+			}),
+
+			isProd && new CleanWebpackPlugin(['build']),
+			isProd && new webpack.optimize.SplitChunksPlugin({}),
+			isProd && new MiniCssExtractPlugin({}),
```

Comment:
Aren’t our CSS files separate already? What is the job of this plugin?

Reply:
Webpack doesn't output CSS by default, instead it actually embeds it into the JS along with bootstrapping code that injects it as a Data URI into the head of the document.  Webpack 4 at least recognizes CSS as input by default, but [MiniCssExtractPlugin](https://twitter.com/wsokra/status/970253245733113856?lang=en) or ExtractTextWebPlugin needs to be used in order to rip the CSS back out of the bundled JS and produce files on disk.